### PR TITLE
Fix crash when typing in an indexer.

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
Fixes https://devdiv.visualstudio.com/DevDiv/_workitems/edit/1167540

We have a cache that attempts to reuse semantic models if edits are made entirely within a method-block or accessor-block.  The code that dealt with accessors only handled properties/events, forgetting that indexers are a different syntactic form entirely.

### CODE REVIEW

File: src/Workspaces/CSharp/Portable/SemanticModelReuse/CSharpSemanticModelReuseLanguageService.cs [hunk 1]
```diff
@@ -31,11 +32,15 @@ public CSharpSemanticModelReuseLanguageService()
...
+        protected override MemberDeclarationSyntax GetAccessorContainerDeclaration(AccessorDeclarationSyntax currentAccessor)
+        {
+            Contract.ThrowIfFalse(currentAccessor.Parent is AccessorListSyntax);
+            Contract.ThrowIfFalse(currentAccessor.Parent.Parent is MemberDeclarationSyntax);
+            return (MemberDeclarationSyntax)currentAccessor.Parent.Parent;
+        }
 
-        protected override SyntaxList<AccessorDeclarationSyntax> GetAccessors(PropertyDeclarationSyntax property)
-            => property.AccessorList?.Accessors ?? default;
+        protected override SyntaxList<AccessorDeclarationSyntax> GetAccessors(MemberDeclarationSyntax member)
```
Comment (sharwell) [inline src/Workspaces/CSharp/Portable/SemanticModelReuse/CSharpSemanticModelReuseLanguageService.cs:42]: ❔ Why not `BasePropertyDeclarationSyntax`?

File: src/Workspaces/CSharp/Portable/CodeGeneration/CSharpSyntaxGenerator.cs [hunk 2]
```diff
@@ -2494,12 +2494,12 @@ public override SyntaxNode InsertAccessors(SyntaxNode declaration, int index, IE
...
+            => declaration switch
             {
-                SyntaxKind.PropertyDeclaration => ((PropertyDeclarationSyntax)declaration).AccessorList,
-                SyntaxKind.IndexerDeclaration => ((IndexerDeclarationSyntax)declaration).AccessorList,
-                SyntaxKind.EventDeclaration => ((EventDeclarationSyntax)declaration).AccessorList,
+                PropertyDeclarationSyntax property => property.AccessorList,
+                IndexerDeclarationSyntax indexer => indexer.AccessorList,
+                EventDeclarationSyntax @event => @event.AccessorList,
                 _ => null,
             };
```
Comment (sharwell) [inline src/Workspaces/CSharp/Portable/CodeGeneration/CSharpSyntaxGenerator.cs:2504]: 💡 

```
(declaration as BasePropertyDeclarationSyntax)?.AccessorList
```

File: src/Workspaces/CSharp/Portable/SemanticModelReuse/CSharpSemanticModelReuseLanguageService.cs (hunk 1 above)
Comment (CyrusNajmabadi) [reply src/Workspaces/CSharp/Portable/SemanticModelReuse/CSharpSemanticModelReuseLanguageService.cs:42]: taht def works :)

**Comment by CyrusNajmabadi** [comment]
> A couple simplifications can be made using a base type for members with accessor lists

Done.  Thanks for catching taht.  I had totally forgotten we had that common base type.

File: src/Workspaces/CSharp/Portable/SemanticModelReuse/CSharpSemanticModelReuseLanguageService.cs [hunk 3]
```diff
@@ -31,11 +33,15 @@ public CSharpSemanticModelReuseLanguageService()
 
         protected override ISyntaxFacts SyntaxFacts => CSharpSyntaxFacts.Instance;
 
-        protected override SyntaxList<AccessorDeclarationSyntax> GetAccessors(EventDeclarationSyntax @event)
-            => @event.AccessorList?.Accessors ?? default;
+        protected override BasePropertyDeclarationSyntax GetBasePropertyDeclaration(AccessorDeclarationSyntax accessor)
+        {
+            Contract.ThrowIfFalse(accessor.Parent is AccessorListSyntax);
+            Contract.ThrowIfFalse(accessor.Parent.Parent is BasePropertyDeclarationSyntax);
```
[the remark this answers is not in the document]
Comment (CyrusNajmabadi) [reply src/Workspaces/CSharp/Portable/SemanticModelReuse/CSharpSemanticModelReuseLanguageService.cs:39]: it also verifies taht it isn't null.

**Pull request merged**